subprojects/core/src/main/java/org/gradle/api/internal/plugins/ImperativeOnlyPluginTarget.java

```diff
@@ -40,7 +50,38 @@ public ConfigurationTargetIdentifier getConfigurationTargetIdentifier() {
     public void applyImperative(@Nullable String pluginId, Plugin<?> plugin) {
         // TODO validate that the plugin accepts this kind of argument
         Plugin<T> cast = uncheckedCast(plugin);
-        cast.apply(target);
+        try {
+            cast.apply(target);
+        } catch (ClassCastException e) {
+            maybeThrowOnTargetMismatch(plugin);
+            throw e;
+        }
+    }
+
+    private void maybeThrowOnTargetMismatch(Plugin<?> plugin) {
+        Type typeParameter = TypeUtils.getTypeArguments(plugin.getClass(), Plugin.class).get(Plugin.class.getTypeParameters()[0]);
+        if (!(typeParameter instanceof Class<?>)) {
+            return;
+        }
+
+        PluginTargetType actualTargetType = PluginTargetType.from((Class<?>) typeParameter);
+        if (actualTargetType == null || targetType.equals(actualTargetType)) {
+            return;
+        }
+
+        throw problems.getInternalReporter()
+            .throwing(spec -> {
+                String message = String.format(
+                    "The plugin must be applied %s, but was applied %s",
+                    actualTargetType.getApplyTargetDescription(), targetType.getApplyTargetDescription()
+                );
+
+                spec.id("target-type-mismatch", "Unexpected plugin type", GradleCoreProblemGroup.pluginApplication())
+                    .severity(Severity.ERROR)
+                    .withException(new IllegalArgumentException(message))
+                    .contextualLabel(message)
+                    .documentedAt(Documentation.userManual("custom_plugins", "project_vs_settings_vs_init_plugins").toString());
```

Comment:
✅ Just wanted to affirm that this is a good usage of the Problems API.

One signal for me, which I'm trying to push on, is the fact that one needs to make the exception beforehand.
I find this a bit redundant, as we could - if not given - make a pretty descriptive exception message from the context available here.

WDYT?

Reply:
What I found redundant in this case is the need to pass the message both to the exception and as a contextual label. What are the cases when they are expected to be different?

I'm also curious if there are cases when creating a problem from exception does not have `ERROR` severity. Maybe there are cases like this, but would it be better to default to `ERROR` severity *if* the problem is based on the exception?

On the other hand, accepting an exception allows the problem-creator to decide on its type, which might be important/helpful. But also it makes the exception stacktrace correspond better to the origin of the problem, rather than start with frames in some "default internal problem reporter builder".

Having said that, I don't have a clear idea of how I would improve this API now to address all of this without potential downsides. I'd imagine keeping things explicit/redundant as they are now might be better because it's always easier/non-breaking to introduce ergonomic methods later and refactor.